File: python/algorithms/bubble_sort.py
#!/usr/bin/env python


def bubble_sort(iterable):
  """Sorts the iterable.

  Bubble sort is a simple, comparison sorting algorithm that iterates through a
  collection, compares adjacent items, and swaps them if they are in the wrong
  order. Iterations through the collection are repeated until no swaps are
  needed.

  * O(n^2) time complexity
  * O(1) space complexity
  * stable
  * adaptive: O(n) when nearly sorted

  Args:
    iterable: A collection that is iterable.
  Raises:
    TypeError: If iterable is not iterable.
  """
  try:
    _ = iter(iterable)
  except TypeError:
    raise TypeError('\'{}\' object is not iterable'.format(
        type(iterable).__name__))

  length = len(iterable)
  unsorted = True
  j = 1
  while unsorted:
    unsorted = False
    for i in range(length - j):
      if iterable[i] > iterable[i + 1]:
        iterable[i], iterable[i + 1] = iterable[i + 1], iterable[i]
        unsorted = True
    j += 1

File: c/algorithms/bubble_sort.c
#include "bubble_sort.h"

#include <stdbool.h>

static void swap(int* i, int* j) {
  int temp = *i;
  *i = *j;
  *j = temp;
}

void bubble_sort(int array[], size_t length) {
  if (length == 0) {
    return;
  }

  bool unsorted = true;
  size_t j = 1;
  while (unsorted) {
    unsorted = false;
    for (size_t i = 0; i < length - j; ++i) {
      if (array[i] > array[i + 1]) {
        swap(array + i, array + i + 1);
        unsorted = true;
      }
    }
    ++j;
  }
}
